File: csharp/Tdcv2/Generators/SymbolGen.cs
using System.Text;
using Tdcv2.Prng;
using Tdcv2.Unicode;

namespace Tdcv2.Generators;

/// <summary>
/// <c>&lt;gen type="symbol" .../&gt;</c> — characters, drawn one at a time.
/// </summary>
/// <remarks>
/// The pool comes from either a named <c>alphabet</c> or an inline <c>value</c> set, never both. To
/// pick a whole word from a list, <c>&lt;gen type="text"&gt;</c> is the tag; this one works below the
/// level of words.
/// </remarks>
public static class SymbolGen
{
    private const int DefaultLength = 1;
    private const int MaxLength = 1024;

    public static IReadOnlyList<string> Generate(
        IReadOnlyDictionary<string, string> attrs, int count, Sfc32 prng)
    {
        IReadOnlyList<string> chars = ResolveChars(attrs);
        int length = ParseLength(attrs.GetValueOrDefault("length"));

        var result = new List<string>(count);
        for (int i = 0; i < count; i++)
        {
            var value = new StringBuilder();
            for (int pos = 0; pos < length; pos++)
            {
                value.Append(Rand.Pick(prng, chars));
            }

            result.Add(value.ToString());
        }

        return result;
    }

    internal static int ParseLength(string? raw)
    {
        if (raw is null)
        {
            return DefaultLength;
        }

        if (!int.TryParse(raw.Trim(), out int value) || value <= 0 || value > MaxLength)
        {
            throw new ArgumentException(
                $"symbol length must be an integer from 1 to {MaxLength}, got \"{raw}\"");
        }

        return value;
    }

    internal static IReadOnlyList<string> ResolveChars(IReadOnlyDictionary<string, string> attrs)
    {
        string? value = BlankToNull(attrs.GetValueOrDefault("value"));
        string? alphabet = BlankToNull(attrs.GetValueOrDefault("alphabet"));

        if (value is not null && alphabet is not null)
        {
            throw new ArgumentException(
                "symbol generator: use either \"value\" (inline set) or \"alphabet\" (named), not both");
        }

        IReadOnlyList<string> baseChars;
        if (value is not null)
        {
            baseChars = CharSet.Parse(value);
            if (baseChars.Count == 0)
            {
                throw new ArgumentException(
                    $"symbol generator: value \"{value}\" produced an empty character set");
            }
        }
        else if (alphabet is not null)
        {
            baseChars = Alphabets.Chars(alphabet)
                ?? throw new ArgumentException(
                    $"unknown alphabet \"{alphabet}\"; known alphabets: "
                    + string.Join(", ", Alphabets.Names()));
        }
        else
        {
            throw new ArgumentException(
                "symbol generator requires \"value\" (inline set like \"abc\" or \"[a-z]\") or "
                + "\"alphabet\" (named); known alphabets: "
                + string.Join(", ", Alphabets.Names()));
        }

        return ApplyIncludeExclude(
            baseChars, attrs.GetValueOrDefault("include"), attrs.GetValueOrDefault("exclude"));
    }

    /// <summary><c>(base ∪ include) − exclude</c>. Exclude is applied last, so it always has the last word.</summary>
    private static IReadOnlyList<string> ApplyIncludeExclude(
        IReadOnlyList<string> baseChars, string? include, string? exclude)
    {
        bool hasInclude = !string.IsNullOrEmpty(include);
        bool hasExclude = !string.IsNullOrEmpty(exclude);
        if (!hasInclude && !hasExclude)
        {
            return baseChars;
        }

        // First-seen order kept, because the set is indexed by a random draw.
        var order = new List<string>(baseChars);
        var seen = new HashSet<string>(baseChars, StringComparer.Ordinal);
        if (hasInclude)
        {
            foreach (string c in CharSet.Parse(include!))
            {
                if (seen.Add(c))
                {
                    order.Add(c);
                }
            }
        }

        if (hasExclude)
        {
            var drop = new HashSet<string>(CharSet.Parse(exclude!), StringComparer.Ordinal);
            order.RemoveAll(drop.Contains);
        }

        if (order.Count == 0)
        {
            throw new ArgumentException(
                "symbol generator: the character set is empty after applying include/exclude");
        }

        return order;
    }

    private static string? BlankToNull(string? s) => string.IsNullOrEmpty(s) ? null : s;
}

File: java/src/main/java/io/github/nickliapin/tdc/generators/SymbolGen.java
package io.github.nickliapin.tdc.generators;

import io.github.nickliapin.tdc.prng.Prng;
import io.github.nickliapin.tdc.prng.Random;
import io.github.nickliapin.tdc.unicode.Alphabets;
import io.github.nickliapin.tdc.unicode.CharSet;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * {@code <gen type="symbol" .../>} — characters, drawn one at a time.
 *
 * <p>The pool comes from either a named {@code alphabet} or an inline {@code value} set, never
 * both. To pick a whole word from a list, {@code <gen type="text">} is the tag; this one works
 * below the level of words.
 */
public final class SymbolGen {

  private static final int DEFAULT_LENGTH = 1;
  private static final int MAX_LENGTH = 1024;

  private SymbolGen() {}

  public static List<String> generate(Map<String, String> attrs, int count, Prng.Sfc32 prng) {
    List<String> chars = resolveChars(attrs);
    int length = parseLength(attrs.get("length"));
    List<String> out = new ArrayList<>(count);
    for (int i = 0; i < count; i++) {
      StringBuilder value = new StringBuilder();
      for (int pos = 0; pos < length; pos++) {
        value.append(Random.pick(prng, chars));
      }
      out.add(value.toString());
    }
    return out;
  }

  static int parseLength(String raw) {
    if (raw == null) {
      return DEFAULT_LENGTH;
    }
    int value;
    try {
      value = Integer.parseInt(raw.trim());
    } catch (NumberFormatException e) {
      throw new IllegalArgumentException(
          "symbol length must be an integer from 1 to " + MAX_LENGTH + ", got \"" + raw + "\"");
    }
    if (value <= 0 || value > MAX_LENGTH) {
      throw new IllegalArgumentException(
          "symbol length must be an integer from 1 to " + MAX_LENGTH + ", got \"" + raw + "\"");
    }
    return value;
  }

  static List<String> resolveChars(Map<String, String> attrs) {
    String value = blankToNull(attrs.get("value"));
    String alphabet = blankToNull(attrs.get("alphabet"));

    if (value != null && alphabet != null) {
      throw new IllegalArgumentException(
          "symbol generator: use either \"value\" (inline set) or \"alphabet\" (named), not both");
    }

    List<String> base;
    if (value != null) {
      base = CharSet.parse(value);
      if (base.isEmpty()) {
        throw new IllegalArgumentException(
            "symbol generator: value \"" + value + "\" produced an empty character set");
      }
    } else if (alphabet != null) {
      base = Alphabets.chars(alphabet);
      if (base == null) {
        throw new IllegalArgumentException(
            "unknown alphabet \"" + alphabet + "\"; known alphabets: " + String.join(", ", Alphabets.names()));
      }
    } else {
      throw new IllegalArgumentException(
          "symbol generator requires \"value\" (inline set like \"abc\" or \"[a-z]\") or \"alphabet\" (named); "
              + "known alphabets: "
              + String.join(", ", Alphabets.names()));
    }

    return applyIncludeExclude(base, attrs.get("include"), attrs.get("exclude"));
  }

  /** {@code (base ∪ include) − exclude}. Exclude is applied last, so it always has the last word. */
  private static List<String> applyIncludeExclude(
      List<String> base, String include, String exclude) {
    boolean hasInclude = include != null && !include.isEmpty();
    boolean hasExclude = exclude != null && !exclude.isEmpty();
    if (!hasInclude && !hasExclude) {
      return base;
    }

    Set<String> chars = new LinkedHashSet<>(base);
    if (hasInclude) {
      chars.addAll(CharSet.parse(include));
    }
    if (hasExclude) {
      CharSet.parse(exclude).forEach(chars::remove);
    }
    if (chars.isEmpty()) {
      throw new IllegalArgumentException(
          "symbol generator: the character set is empty after applying include/exclude");
    }
    return List.copyOf(chars);
  }

  private static String blankToNull(String s) {
    return s == null || s.isEmpty() ? null : s;
  }
}
